### PR TITLE
[YUNIKORN-49] K8shim should report occupied resources to scheduler core

### PR DESCRIPTION
**Pull request by yangwwei**

### CODE REVIEW

**Comment by yangwwei** [comment]
This one is not ready for review yet..

File: pkg/common/resource.go [hunk 1]
```diff
@@ -244,3 +268,35 @@ func Add(left *si.Resource, right *si.Resource) *si.Resource {
...
+	// neither are nil, clone one and sub the other
+	rb := NewResourceBuilder()
+	for k, v := range left.Resources {
+		if rightValue, ok := right.Resources[k]; ok {
+			rb.AddResource(k, v.Value - rightValue.Value)
+		} else {
+			rb.AddResource(k, v.Value)
+		}
+	}
+	return rb.Build()
```
Comment (wilfred-s) [inline pkg/common/resource.go]: if *right* contains resources *left* does not have we do not subtract them.
This is contrary to what we do in add where we properly merge the two.

We should behave the same way in add and sub

File: pkg/common/resource_test.go [hunk 2]
```diff
@@ -243,3 +244,76 @@ func TestNodeResource(t *testing.T) {
...
+		AddResource("a", 1).
+		AddResource("c", 0).
+		AddResource("d", -1).
+		Build()
+	res3 := Sub(res1, res2)
+
+	expected := NewResourceBuilder().
+		AddResource("a", -1).
+		AddResource("b", 1).
+		Build()
```
Comment (wilfred-s) [inline pkg/common/resource_test.go]: That is not what I would expect based on the way we do *add*, see comment in implementation code

File: pkg/cache/node_coordinator.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/common/utils"
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/log"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	k8sCache "k8s.io/client-go/tools/cache"
+)
+
+// nodeResourceCoordinator looks at the resources that not allocated by yunikorn,
+// and refresh scheduler cache to keep nodes' capacity in-sync.
+// this coordinator only looks after the pods that not scheduled by yunikorn,
```
[the remark this answers is not in the document]
Comment (yangwwei) [reply pkg/cache/node_coordinator.go]: done

File: pkg/cache/node_coordinator.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+			return
+		}
+	default:
+		log.Logger.Error("cannot convert to pod")
+		return
+	}
+
+	// if pod is already terminated, that means the updates have already done
+	if utils.IsPodTerminated(pod) {
+		log.Logger.Debug("pod is already terminated, occupied resource updated should have already done")
```
[the remark this answers is not in the document]
Comment (yangwwei) [reply pkg/cache/node_coordinator.go]: done

File: pkg/cache/node_coordinator.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+		return
+	}
+
+	log.Logger.Info("delete pod that scheduled by other schedulers",
```
[the remark this answers is not in the document]
Comment (yangwwei) [reply pkg/cache/node_coordinator.go]: done

File: pkg/cache/task.go [hunk 6]
```diff
@@ -342,19 +345,36 @@ func (task *Task) postTaskCompleted(event *fsm.Event) {
...
 func (task *Task) releaseAllocation() {
 	// when task is completed, we notify the scheduler to release allocations
-	go func() {
-		// scheduler api might be nil in some tests
-		if task.context.apiProvider.GetAPIs().SchedulerAPI != nil {
-			releaseRequest := common.CreateReleaseAllocationRequestForTask(
+	// scheduler api might be nil in some tests
+	if task.context.apiProvider.GetAPIs().SchedulerAPI != nil {
+		// if allocated, sending release
+
```
[the remark this answers is not in the document]
Comment (yangwwei) [reply pkg/cache/task.go]: yeah, removed it. it was the old comment

File: pkg/common/resource.go [hunk 7]
```diff
@@ -244,3 +268,35 @@ func Add(left *si.Resource, right *si.Resource) *si.Resource {
 	}
 	return result
 }
+
+func Sub(left *si.Resource, right *si.Resource) *si.Resource {
+	if left == nil {
+		left = &si.Resource{}
+	}
+	if right == nil {
+		return &si.Resource{}
```
[the remark this answers is not in the document]
Comment (yangwwei) [reply pkg/common/resource.go]: fixed

**Comment by wilfred-s** [comment]
thank you, changes committed

**Pull request closed without merging**